### PR TITLE
fixes #3475 - make it possible to force the 401 status.

### PR DESCRIPTION
**Pull request by adelton**
With this patch, it will be possible to specify

```
ErrorDocument 401 /users/login?status=401
```

to force the preservation of the 401 status.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -70,7 +70,11 @@ def login
         login_user(user)
       end
     else
-      render :layout => 'login'
+      if params[:status]
```
Comment (ares) [inline app/controllers/users_controller.rb]: how about

``` ruby
render :layout => 'login', :status => params[:status] || :ok
```

Comment (adelton) [inline app/controllers/users_controller.rb]: Fine by me, even if you do yet more hardcoding of that :ok (200) response here.

**Comment by adelton** [comment]
Rebased on current develop and used code from https://github.com/theforeman/foreman/pull/1253#discussion_r10074384 -> 07ecb0369fe54bcd04ff45ab4434d064be0f204a.

**Comment by domcleal** [comment]
@adelton where does the status parameter come from in this case?  Are you changing ErrorDocument to redirect to `/user/login?status=401` or similar?

**Comment by adelton** [comment]
Yes, that is the plan -- if thin/rails/foreman do not preserve the status (or rather, force 200), this is a way to be able to force 401 from the ErrorDocument. As a result, the failed Kerberos authentication will be faster because there will be no round-trip to the HTTP client.

**Comment by domcleal** [comment]
Okay, that's fine.  One possible concern came to me thinking about this - are there any dangers to users from reflection attacks?  I think the answer is no, but would be interested in your thought.

e.g. a user is redirected by a malicious site to /users/login?status=123, which causes Foreman to render the page with a "123" status code.  It seems you can get it to return anything from a 0 return code to huge numbers, which Chrome interprets as 2^31!  Am I being overly paranoid - could there be any malicious uses for this?

**Comment by adelton** [comment]
Checking things like

```
# curl -Lksi https://$( hostname )/users/login?status=800000 | grep HTTP/1
```

returns

```
HTTP/1.1 500 Internal Server Error
```

so it looks like the status code is not passed to output verbatim without checking. But I agree that a list of allowed "override" status codes (currently consisting of 401) could be justified. So how about going back to the original change https://github.com/adelton/foreman/commit/53db7544cb3c02a997b4de7b4afda8c1181dcffe and just amending it with explicit check for 401?

``` diff
diff --git a/app/controllers/users_controller.rb b/app/controllers/users_controller.rb
index afc4267..793b716 100644
--- a/app/controllers/users_controller.rb
+++ b/app/controllers/users_controller.rb
@@ -70,7 +70,11 @@ class UsersController < ApplicationController
         login_user(user)
       end
     else
-      render :layout => 'login'
+      if params[:status] && params[:status] == 401
+        render :layout => 'login', :status => params[:status]
+      else
+        render :layout => 'login'
+      end
     end
   end
```

**Comment by domcleal** [comment]
Ah, perhaps I was seeing a different behaviour due to development mode, or because I'm on WEBrick.  A filter for 401 would be best, thanks.

**Comment by adelton** [comment]
OK, please consider the d0526a7351007829cfe181211b3b13becd6b3900 with has the test for 401.

**Comment by domcleal** [comment]
Merged as 5eb2fa6 for Foreman 1.4.2, thanks @adelton.

**Pull request closed without merging**